Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/trace.py

```diff
@@ -0,0 +1,276 @@
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import re
+import sys
+
+from .command import Command
+from webkitscmpy import Commit, local, log, remote
+from webkitbugspy import Tracker
+
+
+COMMIT_REF_BASE = r'r?R?[a-f0-9A-F]+.?\d*@?[0-9a-zA-z\-/]*'
+COMPOUND_COMMIT_REF = r'(?P<primary>{})(?P<secondary> \({}\))?'.format(COMMIT_REF_BASE, COMMIT_REF_BASE)
+CHERRY_PICK_RE = re.compile(r'[Cc]herry[- ][Pp]ick {}'.format(COMPOUND_COMMIT_REF))
+REVERT_RE = re.compile(r'Reverts? {}'.format(COMPOUND_COMMIT_REF))
+FOLLOW_UP_FIXES_RE = [
+    re.compile(r'Fix following {}'.format(COMPOUND_COMMIT_REF)),
+    re.compile(r'Follow-? ?up fix {}'.format(COMPOUND_COMMIT_REF)),
+    re.compile(r'Follow-? ?up {}'.format(COMPOUND_COMMIT_REF)),
+    re.compile(r'\[?[Gg]ardening\]?:? REGRESSION \(?{}\)?'.format(COMPOUND_COMMIT_REF)),
+    re.compile(r'REGRESSION ?\(?{}\)?'.format(COMPOUND_COMMIT_REF)),
+    re.compile(r'\[?[Gg]ardening\]?:? [Tt]est-? ?[Aa]ddition \(?{}\)?'.format(COMPOUND_COMMIT_REF)),
+    re.compile(r'[Tt]est-? ?[Aa]ddition \(?{}\)?'.format(COMPOUND_COMMIT_REF)),
+]
+UNPACK_SECONDARY_RE = re.compile(' \(({})\)'.format(COMMIT_REF_BASE))
+
+
+class Relationship(object):
+    TYPES = (
+        'references', 'referenced by',
+        'cherry-picked', 'original',
+        'reverts', 'reverted by',
+        'follow-up', 'follow-up by',
+    )
+    REFERENCES, REFERENCED_BY, \
+    CHERRY_PICK, ORIGINAL, \
+    REVERTS, REVERTED_BY, \
+    FOLLOW_UP, FOLLOW_UP_BY = TYPES
+
+    PAIRED = [FOLLOW_UP, FOLLOW_UP_BY]
+    IDENTITY = [CHERRY_PICK, ORIGINAL]
+    UNDO = [REVERTS, REVERTED_BY]
+
+    @classmethod
+    def reversed(cls, type):
+        return {
+            cls.REFERENCES: cls.REFERENCED_BY,
+            cls.REFERENCED_BY: cls.REFERENCES,
+            cls.CHERRY_PICK: cls.ORIGINAL,
+            cls.ORIGINAL: cls.CHERRY_PICK,
+            cls.REVERTS: cls.REVERTED_BY,
+            cls.REVERTED_BY: cls.REVERTS,
+            cls.FOLLOW_UP: cls.FOLLOW_UP_BY,
+            cls.FOLLOW_UP_BY: cls.FOLLOW_UP,
+        }.get(type, type)
+
+    @classmethod
+    def parse(cls, commit):
+        lines = commit.message.splitlines()
+
+        for type, regexes in {
+            cls.ORIGINAL: [CHERRY_PICK_RE],
+            cls.REVERTS: [REVERT_RE],
+            cls.FOLLOW_UP: FOLLOW_UP_FIXES_RE,
```

Comment:
Why does `FOLLOW_UP_FIXES_RE` not match the `list` pattern of `CHERRY_PICK_RE` and `REVERT_RE`?